File: leetcode/Jump_Game_II.java
import java.util.*;

public class Solution {
    public int jump(int[] A) {
        int[] d = new int[A.length];
        int p = 0;
        Arrays.fill(d, Integer.MAX_VALUE);
        d[0] = 0;
        for (int i = 0; i < A.length; ++i)
            if (i <= p) {
                int q = Math.min(Math.max(p, i + A[i]), A.length - 1);
                for (int j = p + 1; j <= q; ++j) d[j] = d[i] + 1;
                p = q;
            } else {
                break;
            }
        return d[d.length - 1];
    }
}

File: leetcode/Jump_Game_II.cpp
class Solution {
public:
    int jump(int A[], int n) {
        if (n < 2)
            return 0;
        vector<int> num_jump(n, -1);
        num_jump[0] = 0;
        int rightmost = 0;
        for (int i = 0; i < n - 1; ++i)
            if (rightmost < i + A[i]) {
                for (; rightmost + 1 <= i + A[i] && rightmost + 1 < n;)
                    num_jump[++rightmost] = num_jump[i] + 1;
                if (rightmost == n - 1)
                    return num_jump[n - 1];
            }
        return -1;
    }
};
